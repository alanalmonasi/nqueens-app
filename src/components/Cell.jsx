const queenURL = 'img/queen.png'

const Cell = ({ queen, color }) => {

  const blackBox = parseInt(color);

  const box = {
    width: '100px',
    height: '100px',
    border: '1px black solid',
    backgroundColor: `${(blackBox) % 2 === 0 ? 'white' : 'black'}`
  }

  return (
   <>
    <div style={box} className={blackBox}>{
      queen ? <img src={queenURL} width='100px' height='100px'/> : ''
    }</div>
   </>
  )
}

export default Cell
