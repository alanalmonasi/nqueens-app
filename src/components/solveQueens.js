import { useEffect, useState } from "react";

export const solveQueens = (n) => {
   const nQueens = (n) => {
      let result = [];

      const queens = Array(n).fill(0);

      function makeBoard() {
         const board = [];

         for (let r = 0; r < n; r++) {
            let row = [];

            for (let col = 0; col < n; col++) {
               if (queens[r] === col) {
                  row.push(1);
               } else {
                  row.push(0);
               }
            }
            board.push(row);
         }
         return board;
      }

      function validPos(row, col) {
         for (let i = 0; i < row; i++) {
            if (
               Math.abs(i - row) == Math.abs(queens[i] - col) ||
               queens[i] == col
            ) {
               return false;
            }
         }
         return true;
      }

      function solve(row) {
         if (row === n) {
            const board = makeBoard();
            result.push(board);
         }

         for (let col = 0; col < n; col++) {
            if (!validPos(row, col)) {
               continue;
            }
            queens[row] = col;

            solve(row + 1);

            queens[row] = -1;
         }
      }
      solve(0);
      return result;
   };

   const [chessBoard, setChessBoard] = useState(nQueens(n));

   useEffect(() => {
      setChessBoard(nQueens(n));
   }, [n]);

   return {
      chessBoard,
   };
};
