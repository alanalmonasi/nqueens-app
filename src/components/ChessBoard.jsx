import { useEffect, useState } from "react";
import Cell from "./Cell";
import { solveQueens } from "./solveQueens";
import axios from "axios";

export const ChessBoard = ({ size }) => {
   const chessBox = {
      width: "max-content",
      gridTemplateColumns: `repeat(${size}, 1fr)`,
      display: "grid",
      marginTop: "20px",
      boxShadow: "12px 12px 3px rgba(0,0,0,0.9)",
   };

   const [nextResult, setNextResult] = useState(0);

   const increment = () => {
      if (nextResult === chessBoard.length - 1) return;
      setNextResult(nextResult + 1);
      saveSolution();
   };

   const { chessBoard } = solveQueens(size);

   const currentChessBoard = chessBoard[nextResult];

   const saveSolution = () => {
      const newSolution = {
         sol: currentChessBoard,
         solved: true,
      };
      axios.post("http://localhost:5000/", newSolution);
   };

   //Reset the iteration of results if size changes
   const reset = () => {
      setNextResult(0);
   };

   useEffect(() => {
      reset();
   }, [size]);

   return (
      <>
         <h2>Chessboard</h2>
         <div style={chessBox}>
            {currentChessBoard.map((r, x) =>
               r.map((c, y) => (
                  <Cell queen={c} key={`${x}${y}`} color={x + y} />
               ))
            )}
         </div>
         <div
            style={{ display: size <= 7 ? "none" : "" }}
            className="solutions"
         >
            <button onClick={increment} className="btn btn-primary mt-4">
               Next solution
            </button>
            <hr />
            <div>
               Current solution: {JSON.stringify(currentChessBoard).split(",")}
            </div>
            <div aria-label="total-solutions">
               Total solutions for this board size: {chessBoard.length}
            </div>
            <div id="remaining-solutions">
               Remaining solutions: {chessBoard.length - 1 - nextResult}
            </div>
         </div>
      </>
   );
};
