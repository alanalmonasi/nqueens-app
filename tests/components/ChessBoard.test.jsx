import { render, screen, fireEvent } from "@testing-library/react";
import { ChessBoard } from "../../src/components/ChessBoard";

describe("testing ChessBoard component", () => {
   test("should show default component", () => {
      const { container } = render(<ChessBoard size={0} />);
      expect(container).toMatchSnapshot();
   });

   test("Number of solutions should be 2 if board size is 4", () => {
      render(<ChessBoard size={4} />);
      expect(screen.getByLabelText("total-solutions").innerHTML).toContain(
         "Total solutions for this board size: 2"
      );
   });

   test("Number of solutions should be 92 if board size is 8", () => {
      //if you put other size it will give error and tell the number of solutions for that size
      render(<ChessBoard size={8} />);
      expect(screen.getByLabelText("total-solutions").innerHTML).toContain(
         "92"
      );
   });

   test("Number of queens on board should be the same of size number", () => {
      const n = 9;
      render(<ChessBoard size={n} />);
      const img = screen.getAllByRole("img").length;
      expect(img).toBe(n);
   });
});
