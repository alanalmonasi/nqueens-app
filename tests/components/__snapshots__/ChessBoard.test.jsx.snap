// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ChessBoard component should show default component 1`] = `
<div>
  <h2>
    Chessboard
  </h2>
  <div
    style="grid-template-columns: repeat(0, 1fr); display: grid; margin-top: 20px; box-shadow: 12px 12px 3px rgba(0,0,0,0.9);"
  />
  <div
    class="solutions"
    style="display: none;"
  >
    <button
      class="btn btn-primary mt-4"
    >
      Next solution
    </button>
    <hr />
    <div>
      Current solution: 
      []
    </div>
    <div
      aria-label="total-solutions"
    >
      Total solutions for this board size: 
      1
    </div>
    <div
      id="remaining-solutions"
    >
      Remaining solutions: 
      0
    </div>
  </div>
</div>
`;
